contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -81,17 +79,98 @@ contract KeystoneForwarder is IForwarder, OwnerIsCreator, ITypeAndVersion {
     bool result
   );
 
-  string public constant override typeAndVersion = "KeystoneForwarder 1.0.0";
+  string public constant override typeAndVersion = "Forwarder and Router 1.0.0";
 
-  constructor(address router) OwnerIsCreator() {
-    s_router = router;
+  constructor() OwnerIsCreator() {
+    s_forwarders[address(this)] = true;
   }
 
   uint256 internal constant MAX_ORACLES = 31;
   uint256 internal constant METADATA_LENGTH = 109;
   uint256 internal constant FORWARDER_METADATA_LENGTH = 45;
   uint256 internal constant SIGNATURE_LENGTH = 65;
 
+  // ================================================================
+  // │                          IRouter                             │
+  // ================================================================
+
+  mapping(address forwarder => bool) internal s_forwarders;
+  mapping(bytes32 transmissionId => TransmissionInfo) internal s_transmissions;
+
+  function addForwarder(address forwarder) external onlyOwner {
+    s_forwarders[forwarder] = true;
+    emit ForwarderAdded(forwarder);
+  }
+
+  function removeForwarder(address forwarder) external onlyOwner {
+    s_forwarders[forwarder] = false;
+    emit ForwarderRemoved(forwarder);
+  }
+
+  function route(
+    bytes32 transmissionId,
+    address transmitter,
+    address receiver,
+    bytes calldata metadata,
+    bytes calldata validatedReport
+  ) public returns (bool) {
+    if (!s_forwarders[msg.sender]) {
```

Comment:
Does this work? Isn't the sender a node's transmitter address now?

Also, when we need to start using an new, external forwarder, is there a way to block this one?

Reply:
> Does this work? Isn't the sender a node's transmitter address now?

No, this is because we do `this.route` below, which makes it so that the contract becomes the sender.

> when we need to start using an new, external forwarder, is there a way to block this one?

Yes, we can remove it from the list of forwarders.